Emit a warning when converting to datetime or timedelta values to nanosecond precision

spencerkclark opened the pull request:
This PR addresses #7175 by converting datetime or timedelta values to nanosecond precision even if pandas does not.  For the time being we emit a warning when pandas does not do the conversion, but we do (right now this is only in the development version of pandas).  When things stabilize in pandas we can consider relaxing this constraint in xarray as well.

This got a little bit more complicated due to the presence of timezone-aware datetimes in pandas, but hopefully the tests cover those cases now.

- [ ] Closes #7175
- [ ] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

Illviljan commented on xarray/tests/test_variable.py:
```diff
@@ -2520,6 +2522,26 @@ def test_datetime(self):
         assert np.dtype("datetime64[ns]") == actual.dtype
 
+    @requires_pandas_version_two
+    def test_tz_datetime(self):
```
```suggestion
    def test_tz_datetime(self) -> None:
```

spencerkclark replied:
Ah indeed.  Thanks for the reminder about these!

Illviljan commented on xarray/tests/test_variable.py:
```diff
@@ -2790,3 +2812,78 @@ def test_from_pint_wrapping_dask(self, Var):
+    ],
+    ids=lambda x: f"{x}",
+)
+def test_datetime_conversion_warning(values, warns_under_pandas_version_two):
```
```suggestion
def test_datetime_conversion_warning(values, warns_under_pandas_version_two) -> None:
```

Illviljan commented on xarray/tests/test_variable.py:
```diff
@@ -2790,3 +2812,78 @@ def test_from_pint_wrapping_dask(self, Var):
+
+
+@requires_pandas_version_two
+def test_pandas_two_only_datetime_conversion_warnings():
```
```suggestion
def test_pandas_two_only_datetime_conversion_warnings() -> None:
```

Illviljan commented on xarray/tests/test_variable.py:
```diff
@@ -2790,3 +2812,78 @@ def test_from_pint_wrapping_dask(self, Var):
+    ],
+    ids=lambda x: f"{x}",
+)
+def test_timedelta_conversion_warning(values, warns_under_pandas_version_two):
```
```suggestion
def test_timedelta_conversion_warning(values, warns_under_pandas_version_two) -> None:
```

dcherian commented:
Thanks @spencerkclark I didn't look at the tests too closely but they seem very thorough!

spencerkclark commented:
Thanks @dcherian -- I did my best to try and cover our bases.  For good measure I also added some assertions that the dtypes after conversion were all of nanosecond precision.

spencerkclark commented:
> For good measure I also added some assertions that the dtypes after conversion were all of nanosecond precision.

...which turned up a case that was not valid under the existing pandas version, so I added another separate test.
